src/libraries/System.Security.Cryptography.Pkcs/src/System/Security/Cryptography/Pkcs/CmsSignature.SlhDsa.cs

```diff
@@ -0,0 +1,131 @@
+            lookup.Add(Oids.SlhDsaSha2_256s, new SlhDsaCmsSignature(Oids.SlhDsaSha2_256s));
+            lookup.Add(Oids.SlhDsaShake256s, new SlhDsaCmsSignature(Oids.SlhDsaShake256s));
+            lookup.Add(Oids.SlhDsaSha2_256f, new SlhDsaCmsSignature(Oids.SlhDsaSha2_256f));
+            lookup.Add(Oids.SlhDsaShake256f, new SlhDsaCmsSignature(Oids.SlhDsaShake256f));
+        }
+
+        private sealed class SlhDsaCmsSignature : CmsSignature
+        {
+            private string _signatureAlgorithm;
+
+            internal SlhDsaCmsSignature(string signatureAlgorithm)
+            {
+                _signatureAlgorithm = signatureAlgorithm;
+            }
+
+            protected override bool VerifyKeyType(object key) => key is SlhDsa;
+            internal override bool NeedsHashedMessage => false;
+
+            internal override RSASignaturePadding? SignaturePadding => null;
+
+            internal override bool VerifySignature(
+#if NET || NETSTANDARD2_1
+                ReadOnlySpan<byte> valueHash,
+                ReadOnlyMemory<byte> signature,
+#else
+                byte[] valueHash,
+                byte[] signature,
+#endif
+                string? digestAlgorithmOid,
+                ReadOnlyMemory<byte>? signatureParameters,
+                X509Certificate2 certificate)
+            {
+                if (signatureParameters.HasValue)
+                {
+                    throw new CryptographicException(
+                        SR.Format(SR.Cryptography_UnknownAlgorithmIdentifier, _signatureAlgorithm));
+                }
+
+                // The spec (as of May 5, 2025) has strength requirements on the hash, but we will
+                // not enforce them here. If the callers wants to enforce them, they can do so by themselves.
+
+                SlhDsa? publicKey = certificate.GetSlhDsaPublicKey();
+
+                if (publicKey is null)
+                {
+                    return false;
+                }
+
+                return publicKey.VerifyData(
+                    valueHash,
+#if NET || NETSTANDARD2_1
+                    signature.Span
+#else
+                    signature
+#endif
+                );
+            }
+
+            protected override bool Sign(
+#if NET || NETSTANDARD2_1
+                ReadOnlySpan<byte> dataHash,
+#else
+                byte[] dataHash,
+#endif
+                string? hashAlgorithmOid,
+                X509Certificate2 certificate,
+                object? key,
+                bool silent,
+                [NotNullWhen(true)] out string? signatureAlgorithm,
+                [NotNullWhen(true)] out byte[]? signatureValue,
+                out byte[]? signatureParameters)
+            {
+                signatureParameters = null;
+
+                // If there's no private key, fall back to the public key for a "no private key" exception.
+                SlhDsa? signingKey = key as SlhDsa ??
```

Comment:
Likewise, more algorithms need to be disposed (and cleanup for the others can happen separately)

Reply:
Just have to be careful here to not dispose of the key if it was passed in externally.

Reply:
> // If there's no private key, fall back to the public key for a "no private key" exception.

Seems like a roundabout way of doing it... can we just throw here ourselves or do we prefer the exception coming from the cert?

Reply:
Since this has a type forward to .NET Framework, we try to be exception compatible with .NET Framework, and this is what .NET Framework does.